src/pdb/mod.rs

```diff
@@ -346,106 +377,150 @@ impl Page {
     /// bits in `row_present_flags`. This instance allows the row itself to be lazily loaded, unless it
     /// is not present, in which case there is no content to be loaded.
-    rows: [Option<FilePtr16<Row>>; Self::MAX_ROW_COUNT],
+    // rustc doesn't seem to recognize that this is used below, ignore for now
+    #[allow(dead_code)]
+    row_offsets: [u16; Self::MAX_ROW_COUNT],
     row_presence_flags: u16,
     /// Unknown field, probably padding.
     ///
     /// Apparently this is not always zero, so it might also be something different.
     unknown: u16,
+    // build rows from offsets collected above
+    #[br(seek_before=SeekFrom::Start(page_heap_position))]
+    #[br(args(page_type))]
+    #[br(parse_with = binrw::file_ptr::parse_from_iter(Self::present_rows_offsets(&row_offsets, row_presence_flags)))]
+    #[br(restore_position)] // ensure the parser points to just after this instance, this is important
+    /// Access rows in this RowGroup
+    rows: Vec<Row>,
 }
 
 impl RowGroup {
     const MAX_ROW_COUNT: usize = 16;
+    const BINARY_SIZE: u32 = (Self::MAX_ROW_COUNT as u32) * 2 + 4; // size the serialized structure
 
-    /// Return the ordered list of row offsets that are actually present.
-    pub fn present_rows(&self) -> impl Iterator<Item = Row> + '_ {
-        self.rows
+    /// Get all rows present in this rowgroup
+    #[must_use]
+    pub fn present_rows(&self) -> &[Row] {
+        &self.rows
+    }
+    // TODO(Swiftb0y): Add a new error category for user APIs and add the correct
+    // error herer
+    #[allow(clippy::result_unit_err)]
+    /// Add a row to this rowgroup
+    pub fn add_row(&mut self, row: Row) -> Result<(), ()> {
+        if self.rows.len() >= Self::MAX_ROW_COUNT {
+            return Err(());
+        }
+        self.row_presence_flags |= 1 << self.rows.len() as u16;
+        self.rows.push(row);
+        Ok(())
+    }
+
+    fn present_rows_offsets(
+        row_offsets: &[u16; Self::MAX_ROW_COUNT],
+        row_presence_flags: u16,
+    ) -> impl Iterator<Item = u16> + '_ {
+        row_offsets
             .iter()
             .rev()
-            .filter_map(|row_offset| row_offset.as_ref().map(|r| r.value.clone()))
+            .enumerate()
+            .filter_map(move |(i, row_offset)| {
+                (row_presence_flags & (1 << i) != 0).then_some(*row_offset)
+            })
     }
 }
 
-impl BinRead for RowGroup {
-    type Args<'a> = (PageType, u64);
+impl PartialEq for RowGroup {
+    fn eq(&self, other: &Self) -> bool {
+        self.rows == other.rows
+    }
+}
 
-    /// Read a row group from the reader.
-    ///
-    /// Note: For this to work, the read position needs to be at the *end* of the row group. For
-    /// the first row group, this is the end of the page heap.
-    fn read_options<R: Read + Seek>(
-        reader: &mut R,
+impl BinWrite for RowGroup {
+    type Args<'a> = (u64, u64);
+
+    fn write_options<W: Write + Seek>(
+        &self,
```

Comment:
this can also be replaced by `write_with` that just delegates to `write_options_and_get_row_offset`